token/program-2022/src/processor.rs

```diff
@@ -352,47 +353,71 @@ impl Processor {
         // compute costs, ie:
         // if self_transfer || amount == 0
         check_program_account(source_account_info.owner)?;
-        check_program_account(dest_account_info.owner)?;
 
         // This check MUST occur just before the amounts are manipulated
         // to ensure self-transfers are fully validated
         if self_transfer {
             return Ok(());
         }
 
-        // self-transfer was dealt with earlier, so this *should* be safe
-        let mut dest_account_data = dest_account_info.data.borrow_mut();
-        let mut dest_account = StateWithExtensionsMut::<Account>::unpack(&mut dest_account_data)?;
-
-        if dest_account.base.is_frozen() {
-            return Err(TokenError::AccountFrozen.into());
-        }
-        if !cmp_pubkeys(&source_account.base.mint, &dest_account.base.mint) {
-            return Err(TokenError::MintMismatch.into());
-        }
+        let wsol_source = source_account.base.is_native();
+        let sol_destination = *dest_account_info.owner == system_program::id();
+        match (wsol_source, sol_destination) {
+            // WSOL -> SOL
+            (true, true) => {
+                if fee > 0 {
+                    return Err(TokenError::NativeNotSupported.into());
+                }
+                source_account.base.amount = source_account
+                    .base
+                    .amount
+                    .checked_sub(amount)
+                    .ok_or(TokenError::Overflow)?;
+            }
+            // SPL -> SOL
+            (false, true) => {
+                return Err(TokenError::MintMismatch.into());
+            }
+            // WSOL -> WSOL | SPL -> SPL
+            (true, false) | (false, false) => {
+                // self-transfer was dealt with earlier, so this *should* be safe
+                check_program_account(dest_account_info.owner)?;
+                let mut dest_account_data = dest_account_info.data.borrow_mut();
+                let mut dest_account =
+                    StateWithExtensionsMut::<Account>::unpack(&mut dest_account_data)?;
+
+                if dest_account.base.is_frozen() {
+                    return Err(TokenError::AccountFrozen.into());
+                }
+                if !cmp_pubkeys(&source_account.base.mint, &dest_account.base.mint) {
+                    return Err(TokenError::MintMismatch.into());
+                }
```

Comment:
Since there's already a wSOL specific instruction, `SyncNative`, would it be simpler if there was another wSOL specific instruction, `Unwrap(amount)` instead of doing this in `Transfer`?